### DISCUSSION TITLE
Firefox notifications shown as full tiles

### DISCUSSION THREAD

**Discussion by git1106** [i3/i3#5966] [Q&A] [answered]
Firefox have changed their built-in notifications (turned on with `about:config` `alerts.useSystemBackend=false`) several times recently. Now notifications have appeared as fully tiling windows in the 2 latest releases.

It had always worked before with `i3` so that notification where just little popups in the upper right corner.
 
I reported it in https://bugzilla.mozilla.org/show_bug.cgi?id=1881434 , please find screenshots and `xprop` output there.

Can anybody tell from that data what is wrong? Or give hints how to debug the issue?

I also built the latest `i3` release from source, but it behaves exactly the same way.

**Comment by thecrawler1** [comment]
I fixed with `for_window [window_role="alert"] floating enable`

**Reply by git1106** [reply]
Thanks, that helps. But why is that needed? Not every user should have to figure that out, neither themselves nor by asking in a forum. I was nearly sure this can be influenced by configuration, but I did not even search because I am convinced it should work sensibly out of the box.

This change could have much wider impact than Firefox notifications, I don't know who else uses role `alert`.

* Does Firefox do something wrong nowadays?
* Are the `i3` defaults not sensible?
* Something else?

**Reply by TyberiusPrime** [reply]
For me that makes the widows float correctly, but they still steal focus.
I need to add `no_focus [window_role="alert"]` as well to fix that.

**Comment by KebabLord** [comment]


**Reply by git1106** [reply]
There is no doubt that notifications displayed by other programs are displayed correctly. Older versions of `firefox` still do, I tested `xfce4-notifyd`, and you tested `dunst`.

But that does not answer the question why current versions of `firefox` don't work (for a default `i3` configuration). Did they do something wrong or have they hit a bug in `i3`?

**Comment by KebabLord** [comment, marked as answer]
I guess figured out the problem, apparently in newer releases of firefox, the notifications window type has changed from:
`WM_WINDOW_TYPE_DIALOG` to `WM_WINDOW_TYPE_NOTIFICATION` (just like you mentioned in the bugzilla)

i3 Doesn't mark `WINDOW_TYPE_NOTIFICATION` as floating as seen here: 
https://github.com/i3/i3/blob/next/src/manage.c#L466

A quick patch on `manage.c` like this might fix it:

```patch
diff --git a/src/manage.c b/src/manage.c
index b0684d6b..873dc5c3 100644
--- a/src/manage.c
+++ b/src/manage.c
@@ -467,6 +467,7 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
         xcb_reply_contains_atom(type_reply, A__NET_WM_WINDOW_TYPE_UTILITY) ||
         xcb_reply_contains_atom(type_reply, A__NET_WM_WINDOW_TYPE_TOOLBAR) ||
         xcb_reply_contains_atom(type_reply, A__NET_WM_WINDOW_TYPE_SPLASH) ||
+        xcb_reply_contains_atom(type_reply, A__NET_WM_WINDOW_TYPE_NOTIFICATION) ||
         xcb_reply_contains_atom(state_reply, A__NET_WM_STATE_MODAL) ||
         (cwindow->max_width > 0 && cwindow->max_height > 0 &&
          cwindow->min_height == cwindow->max_height &&
```

**Reply by orestisfl** [reply]
Feel free to open a PR for that patch

**Reply by orestisfl** [reply]
PR merged: https://github.com/i3/i3/pull/6017, thank you

**Comment by KebabLord** [comment]
After a bit further digging, i came to conclusion that this problem is caused by firefox, not i3.
Chrome is also using WM_WINDOW_TYPE_NOTIFICATION and i3 successfully sets it as floating.
The reason why firefox popups are not floating is problably because xprops of firefox notification is bad.

On Chrome WM_STATE values, we can see that window is set to "ABOVE EVERYTHING" 
`_NET_WM_STATE(ATOM) = _NET_WM_STATE_ABOVE, _NET_WM_STATE_SKIP_TASKBAR`

Firefox WM_STATE value on the other hand is empty
`_NET_WM_STATE(ATOM) = `

Additionally WM_WINDOW_ROLE value of firefox is "alert" which should been "popup".
I'm not expert on xlib but if chrome is working with same window type while firefox can't, it is not the problem of  i3.

**Reply by git1106** [reply]
Thanks @KebabLord for your digging. I have also noticed that some programs displaying notifications use `_NET_WM_STATE_ABOVE`. That *sounds* like it could be appropriate, but I am not familiar with the exact specification and implementation details. Interestingly enough older Firefox releases got the notification windows displayed in `i3` as small floating windows above everything else **without** that setting. The `xprop` outputs are in the Mozilla bug report linked above.

Unfortunately I might not have time to investigate more during the next days.
